src/buffer_research.py

```diff
@@ -0,0 +1,232 @@
+    df["external_vol_across_time"] = external_vol_across_time
+    df["nr_of_internal_trades"] = nr_of_internal_trades
+    df["nr_of_external_trades"] = nr_of_external_trades
+    return df
+
+
+if __name__ == '__main__':
+
+    fetch_data_from_dune = False
+    verbose_logging = False
+
+    if fetch_data_from_dune:
+        dune_connection = DuneAnalytics.new_from_environment()
+        dune_data = dune_connection.fetch(
+            query_filepath="./src/dune_api/queries/dex_ag_trades_within_one_day.sql",
+            network='mainnet',
+            name="dex ag trades",
+        )
+        with open("data/dune_trading_data_download.txt", "bw+") as f:
+            pickle.dump(dune_data, f, protocol=pickle.HIGHEST_PROTOCOL)
+    else:
+        with open("data/dune_trading_data_download.txt", "rb") as f:
+            dune_data = pickle.load(f)
+
+    df = pd.DataFrame.from_records(dune_data)
+    # We are sorting out all trades that don't have prices. This number of trades is neglegible for the overall analysis
+    df = df[df['usd_amount'].notna()]
+    # scaling factor to scale revenue for 1 year
+    revenue_scaling_factor = 360*24*60*60 / \
+        ((df['block_number'].max() - df['block_number'].min()) * 13.5)
+
+    # preparations for writing simulation results into csv file
+    result_file = 'result.csv'
+    header = ['Buffer Investment [USD]',
+              'AllowListed Buffer-Tokens',
+              'Ratio Internal Trades',
+              'Ratio Internally Matched Vol',
+              'Fee Revenue [USD/Year]',
+              'Gas Saving vs Normal Trade',
+              'Gas Saving(in v2) vs Normal Trade',
+              ]
+    with open(result_file, 'w', encoding='UTF8') as f:
+        writer = csv.writer(f)
+        writer.writerow(header)
+    # gas costs rough estimations
+    gas_cost_avg_amm_trade = 200000
+    gas_cost_over_head_cowswap = 90000
+    gas_cost_internal_trade = 130000
+    gas_cost_over_head_cowswap_in_potential_v2 = 55000
+    gas_cost_internal_trade_in_potential_v2 = 105000
+
+    # Starting simulation with different model parameters
+    for initial_buffer_value_in_usd in [2_000_000, 10_000_000, 50_000_000]:
+        for trade_activity_threshold_for_buffers_to_be_funded in [0.01, 0.001, 0.0005, 0]:
+
+            tokens = set.union({t for t in df['token_a_address']}, {
+                t for t in df['token_b_address']})
+
+            normalized_token_appearance_counts = df['token_b_address'].value_counts(
+                normalize=True)
+            buffer_allow_listed_tokens = list({t for t in tokens if (t in normalized_token_appearance_counts
+                                                                     and normalized_token_appearance_counts[t] > trade_activity_threshold_for_buffers_to_be_funded)})
+            buffers = {t: t in buffer_allow_listed_tokens and initial_buffer_value_in_usd /
+                       len(buffer_allow_listed_tokens) or 0 for t in tokens}
```

Comment:
I still don't get what this dictionary is. I am parsing it as follows:

```python3
{t: (t in some_set) and some_scalar or 0 }
```

The "or 0" part is redundant, because `Anything \/ False = Anything` . I am confused, help!

Reply:
It's ugly python notation: In js I would write:
```
 t in buffer_allow_listed_tokens ? initial_buffer_value_in_usd /
                       len(buffer_allow_listed_tokens) : 0
 ```
 It's a short form for conditional statement